server/logging_middleware.go

```diff
@@ -0,0 +1,114 @@
+package server
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/bmeg/grip/log"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+func extractHeaderKeys(input map[string][]string, whitelist []string) map[string][]string {
+	filtered := make(map[string][]string)
+	if input == nil {
+		return filtered
+	}
+	for k, v := range input {
+		for _, w := range whitelist {
+			if strings.ToLower(k) == strings.ToLower(w) {
+				filtered[strings.ToLower(k)] = v
+				break
+			}
+		}
+	}
+	return filtered
+}
+
+// Return a new interceptor function that logs all requests at the Info level
+func unaryInterceptor(disabled bool, whitelist []string) grpc.UnaryServerInterceptor {
+	// Return a function that is the interceptor.
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if disabled {
+			return handler(ctx, req)
+		}
+		start := time.Now()
+		resp, err := handler(ctx, req)
+		md, _ := metadata.FromIncomingContext(ctx)
+		headers := extractHeaderKeys(md, whitelist)
+		entry := log.WithFields(log.Fields{
+			"path":    info.FullMethod,
+			"request": req,
+			"headers": headers,
+			"latency": time.Since(start).String(),
+			"status":  runtime.HTTPStatusFromCode(status.Code(err)),
+		})
+		if err == nil {
+			entry.Info("gRPC server responded")
+		} else {
+			entry.Error("gRPC server responded")
```

Comment:
Are both paths supposed to have the same message? Maybe log the contents of err?

Reply:
Yeah the only difference was supposed to be the log level. I can add the error message, the reason I didn’t initially was to make these logs as similar to the ones from the http server as possible.